### DISCUSSION TITLE
Installing theseus with `pip install theseus-ai` with a specific version of PyTorch

### DISCUSSION THREAD

**Discussion by DanielTakeshi** [facebookresearch/theseus#530] [Q&A] [answered]
From reading the README, I am trying to install `theseus` using `pip install` but with a specific version of PyTorch (for example PyTorch 1.13.0). For that, I followed these commands:

```
conda create --name theseus python=3.8 -y && conda activate theseus
conda install ipython jupyter -y
conda install pytorch==1.13.0 torchvision==0.14.0 torchaudio==0.13.0 pytorch-cuda=11.6 -c pytorch -c nvidia -y
pip install theseus-ai
```

You can find the raw pastebin with the output here:

https://pastebin.com/raw/jFEXvawV

However I notice that `pip install theseus-ai` will override my older `torch` version to use version 2.0.1, which is the latest version as of today. Is there a reason for this, out of curiosity? is it related to this deprecation of older torch? https://github.com/facebookresearch/theseus/pull/525/files

**Comment by luisenp** [comment, marked as answer]
HI @DanielTakeshi. It's sort of related, but not a consequence of that PR. A few months ago we made this the behavior of our wheels uploaded to pypi, after `torch` 1.13 deprecated some older CUDA versions, and in anticipation of eventually moving to `torch` 2.0. 

If you want a more custom install, you could try installing from source, or also try modifying [this](https://github.com/facebookresearch/theseus/blob/main/build_scripts/build_wheel.sh) script, which is the one we use to build our wheels.  In line 67 you can see the version is set to ">=1.13", unless you are using CUDA 10.2 or 11.3. Changing that line to equality and running the script should generate what you need. You can then run as

```bash
./build_scripts/build_wheel.sh ROOT_DIR 0.1.4 11.6
```

And then pip install from the generated wheel that's saved under ROOT_DIR. 

Let me know if you have any questions.

**Reply by DanielTakeshi** [reply]
Thanks! I think that answers my question.

**Reply by DanielTakeshi** [reply]
Update: just following up that I was able to get this working with PyTorch 1.13. The change that I had to do to the `build_wheel.sh` script was in 2 places (python 3.8 and torch version):

```
(theseus) seita@starship:~/theseus (main) $ git diff
diff --git a/build_scripts/build_wheel.sh b/build_scripts/build_wheel.sh
index c969b31..140013b 100755
--- a/build_scripts/build_wheel.sh
+++ b/build_scripts/build_wheel.sh
@@ -37,7 +37,7 @@ CUDA_VERSION=$3
 TH_VERSION=${4:-${COMMIT}}
 INCLUDE_LABS=${5:-0}
 NIGHTLY=${6:-0}
-PY_VER=${7:-3.10}
+PY_VER=${7:-3.8}
 
 if [[ ${INCLUDE_LABS} != 0 ]]
 then
@@ -66,7 +66,7 @@ CUDA_VERSION_IS_SUPPORTED=$(echo "cpu ${SUPPORTED_CUDA_VERSIONS}" | grep -w ${CU
 
 CUDA_SUFFIX=$(echo ${CUDA_VERSION} | sed 's/[.]//g')
 
-TORCH_VERSION='"torch>=1.13"'
+TORCH_VERSION='"torch==1.13"'
 if [[ ${CUDA_VERSION} == "cpu" ]]
 then 
     DEVICE_TAG=cpu
(theseus) seita@starship:~/theseus (main) $ 
```

The command to build the script:

```
./build_scripts/build_wheel.sh ROOT_DIR 0.2.1 11.6
```

Then create and install stuff:

```
conda create --name theseus python=3.8 -y && conda activate theseus
conda install pytorch==1.13.0 torchvision==0.14.0 torchaudio==0.13.0 pytorch-cuda=11.6 -c pytorch -c nvidia -y
pip install ROOT_DIR/theseus_docker_3.8/theseus_ai-0.2.1+cu116-cp38-cp38-manylinux_2_17_x86_64.whl
```

After that, I can do the following and it works:

```
(theseus) seita@starship:~/theseus (main) $ python
Python 3.8.17 (default, Jul  5 2023, 21:04:15) 
[GCC 11.2.0] :: Anaconda, Inc. on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import torch, theseus
/home/seita/theseus/theseus/_version.py:30: FutureWarning: Using torch < 2.0 for theseus is deprecated and compatibility will be discontinued in future releases.
  warnings.warn(
>>> torch.__version__
'1.13.0'
>>> theseus.__version__
'0.2.1'
>>> exit()
(theseus) seita@starship:~/theseus (main) $ 
```

**Reply by luisenp** [reply]
Thanks for sharing! Would you be interested in submiting a PR adding a CLI flag to specify the torch version to install (e.g. "==1.13", ">=1.13")? The python version can already be specified via CLI, so no need to add one for that.
